atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraRequestExceptionHandler.java

```diff
@@ -0,0 +1,169 @@
+                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+
+            try {
+                Thread.sleep(sleepDuration);
+            } catch (InterruptedException i) {
+                throw new RuntimeException(i);
+            }
+            if (req.getNumberOfAttempts() >= maxTriesSameHost.get()) {
+                req.giveUpOnPreferredHost();
+            }
+        } else if (isFastFailoverException(ex)) {
+            log.info("Retrying with fast failover a query intended for host {}.",
+                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+            req.giveUpOnPreferredHost();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private <K extends Exception> void handleExceptionInternal(
+            RetryableCassandraRequest<?, K> req, InetSocketAddress host, Exception ex) throws K {
+        if (isRetriableException(ex) || isRetriableWithBackoffException(ex) || isFastFailoverException(ex)) {
+            if (req.getNumberOfAttempts() >= maxTriesTotal.get()) {
+                if (ex instanceof TTransportException
+                        && ex.getCause() != null
+                        && (ex.getCause().getClass() == SocketException.class)) {
+                    log.error(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts(), ex);
+                    String errorMsg =
+                            MessageFormatter.format(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts()).getMessage();
+                    throw (K) new TTransportException(((TTransportException) ex).getType(), errorMsg, ex);
+                } else {
+                    log.error("Tried to connect to cassandra {} times.",
+                            SafeArg.of("numTries", req.getNumberOfAttempts()), ex);
+                    throw (K) ex;
+                }
+            } else {
+                // Only log the actual exception the first time
+                if (req.getNumberOfAttempts() > 1) {
+                    log.warn("Error occurred talking to cassandra. Attempt {} of {}. Exception message was: {} : {}",
+                            SafeArg.of("numTries", req.getNumberOfAttempts()),
+                            SafeArg.of("maxTotalTries", maxTriesTotal),
+                            SafeArg.of("exceptionClass", ex.getClass().getTypeName()),
+                            UnsafeArg.of("exceptionMessage", ex.getMessage()));
+                } else {
+                    log.warn("Error occurred talking to cassandra. Attempt {} of {}.",
+                            SafeArg.of("numTries", req.getNumberOfAttempts()),
+                            SafeArg.of("maxTotalTries", maxTriesTotal),
+                            ex);
+                }
+                if (isConnectionException(ex) && req.getNumberOfAttempts() >= maxTriesSameHost.get()) {
+                    blacklist.add(host);
+                }
+            }
+        } else {
+            throw (K) ex;
+        }
+    }
+    @VisibleForTesting
```

Comment:
nit: missing newline between methods.
It's also a bit sad that these need to be `VisibleForTesting`.

Reply:
We could rewrite the tests in `CassandraRequestExceptionHandlerTest` to inject the exception in `handleExceptionInternal`. But since we're changing and configuring some of this logic with this chain of PRs, let's refactor those unit tests later.